google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java

```diff
@@ -29,47 +45,106 @@
 
   @Override
   public Timestamp write(Iterable<Mutation> mutations) throws SpannerException {
-    return pool.getReadWriteSession().write(mutations);
+    Span span = tracer.spanBuilder(READ_WRITE_TRANSACTION).startSpan();
+    try (Scope s = tracer.withSpan(span)) {
+      return pool.getReadWriteSession().write(mutations);
+    } catch (RuntimeException e) {
+      TraceUtil.endSpanWithFailure(span, e);
+      throw e;
+    } finally {
+      span.end();
+    }
   }
 
   @Override
   public Timestamp writeAtLeastOnce(Iterable<Mutation> mutations) throws SpannerException {
-    return pool.getReadSession().writeAtLeastOnce(mutations);
+    Span span = tracer.spanBuilder(READ_WRITE_TRANSACTION).startSpan();
+    try (Scope s = tracer.withSpan(span)) {
+      return pool.getReadSession().writeAtLeastOnce(mutations);
+    } catch (RuntimeException e) {
+      TraceUtil.endSpanWithFailure(span, e);
+      throw e;
+    } finally {
+      span.end();
+    }
   }
 
   @Override
   public ReadContext singleUse() {
-    return pool.getReadSession().singleUse();
+    Span span = tracer.spanBuilder(READ_ONLY_TRANSACTION).startSpan();
+    try (Scope s = tracer.withSpan(span)) {
+      return pool.getReadSession().singleUse();
+    } catch (RuntimeException e) {
+      TraceUtil.endSpanWithFailure(span, e);
+      throw e;
+    }
```

Comment:
You probably need to have a finally here as well. (more cases like this in the rest of the file).